src/lib/support/Pool.h

```diff
@@ -0,0 +1,146 @@
+/*
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *
+ * @brief
+ *   Defines a header only BitMapObjectPool.
+ */
+
+#pragma once
+
+#include <array>
+#include <assert.h>
+#include <atomic>
+#include <limits>
+#include <new>
+#include <stddef.h>
+
+namespace chip {
+
+/**
+ *  @brief
+ *   A class template used for allocating Objects.
+ *
+ *  @tparam     T   a subclass of Object to be allocated.
+ *  @tparam     N   a positive integer max number of objects the pool provides.
+ */
+template <class T, size_t N>
+class BitMapObjectPool
+{
+public:
+    BitMapObjectPool()
+    {
+        assert(std::atomic<tBitChunkType>{}.is_lock_free());
```

Comment:
Why do we want this here?  If we could do it as a `static_assert`, that might be useful, but what does a runtime check buy us?

Reply:
No, can't use `static_assert` here, it is a member function call. I found `std::atomic<T>::is_always_lock_free` but it is only available until C++17

Reply:
I added this check because I want to ensure that atomic implementation isn't using mutex or lock.

The runtime check will opt-out in production code, since assert only affect debugging build.

Reply:
"Production code" apparently typically does NOT define `-NDEBUG` in this space, because they want to get actual assertion failures.

If the atomic implementation _were_ using a mutex, how would we react?  If we just want to be alerted that this code should be changed, I'd rather have a test that tests this, with a pointer to this code in the test...

Reply:
Make sense, I have deleted the check.